### PR TITLE
Insights for Stats

### PR DESCRIPTION
**Pull request by astralbodies**
Adds a new screen for stats in addition to days/weeks/months/years that includes insights into the users' site.

StatsDemo released 4-Jun includes the changes from this PR.

Things being handed off for another PR:
* All Time module 2x2 on portrait iPhone and 1x4 on landscape or iPad
* Today Stats each section tappable & selects graph for that section (Views, Visitors, Likes, Comments)
* Moving Comments, Followers, Tags and Publicize modules into Insights

Needs Review: @jancavan (design) @jleandroperez (code) and maybe @sendhil

### CODE REVIEW

**Comment by jancavan** [comment]
This is looking really good @astralbodies! 

Here's some feedback:

**1 — The all-time numbers font sizes are varied on portrait view:**
4-rows

Is it possible to break it up into 2 rows instead for it to be more scalable?
2-rows

**2 — "Best Ever Views" label wraps on landscape view:**
landscape

On Calypso, the All-time section is broken up into 2 rows for non-English users to make sure labels don't wrap when they are longer. PR: https://github.com/Automattic/calypso-pre-oss/pull/6520

**3 — Views percentage labels need to be updated:**
views

**4 — Months should probably be the full word to avoid i18n issues. So October instead of Oct:**
contract

**5 — "Today's Stats" module heading should be `$blue-wordpress` `(#0087be)` since this is a link:**
link

**6 — Labels above the numbers in Today's Stats are not centered correctly:**
center

**7 — A couple of centering issues on All-time labels as well:**
centering

**8 — When numbers have not loaded yet, let's use single instead of double dashes. The color for this is `lighten( $gray, 20% )` `#c8d7e1`**:

On Calypso, the dash also pulsates if the number has not loaded yet (but that animation can probably just be added in a future iteration).

img_1458

**Comment by astralbodies** [comment]
Thanks for the feedback @jancavan! Few comments on some of the changes:

> 1 — The all-time numbers font sizes are varied on portrait view:
> Is it possible to break it up into 2 rows instead for it to be more scalable?
and
> 2 — "Best Ever Views" label wraps on landscape view:
I have the font in those labels set to reduce by up to 50% to fit the numbers. Unfortunately that autofit is for each label. I think the right solution is to go to 2x2 on each of those two modules. This isn't a trivial amount of work (autolayout in iOS is great but challenging). Logged issue #264 for this and will make it required for the release.

> 4 — Months should probably be the full word to avoid i18n issues. So October instead of Oct
I'm actually using an `NSDateFormatter` (Apple provided date to string thing) so localization won't be an issue. It's what Apple calls "medium style date" format. In English it's three character months but in other languages it'll be more appropriate for their locale.

> 6 — Labels above the numbers in Today's Stats are not centered correctly:
> 7 — A couple of centering issues on All-time labels as well:
This might be due to cropping issues of the images @jasmussen created for me. I'll let you know ASAP.

I'll ping you when the changes are ready for review.

File: WordPressCom-Stats-iOS/InsightsTableViewController.m [hunk 1]
```diff
@@ -0,0 +1,294 @@
+#import "InsightsTableViewController.h"
+#import "WPFontManager+Stats.h"
+#import "WPStyleGuide+Stats.h"
+#import "StatsTableSectionHeaderView.h"
+#import "StatsGaugeView.h"
+
+@interface InlineTextAttachment : NSTextAttachment
+
+@property CGFloat fontDescender;
```
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/InsightsTableViewController.m]: Note here: This property will be handled as *atomic, assign, readwrite*. Do we need this one to be atomic?

File: WordPressCom-Stats-iOS/InsightsTableViewController.m [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+
+
+#pragma mark - UITableViewDelegate methods
+
+- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    [tableView deselectRowAtIndexPath:indexPath animated:YES];
+    
+    if (indexPath.section == 2) {
```
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/InsightsTableViewController.m]: Me being annoying: Maybe a constant?

File: WordPressCom-Stats-iOS/StatsEphemory.m [hunk 3]
```diff
@@ -13,7 +13,7 @@ @implementation EphemoryContainer
 
 @interface StatsEphemory ()
 
-@property (nonatomic, strong) NSCache *cache;
+@property (nonatomic, strong) NSMutableDictionary *cache;
```
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/StatsEphemory.m]: Out of the curiosity, did NSCache bring you any issues?

File: WordPressCom-Stats-iOS/WPStatsService.m [hunk 4]
```diff
@@ -173,6 +204,90 @@ - (void)retrieveAllStatsForDate:(NSDate *)date
...
+                                                 progressBlock:(void (^)(NSUInteger numberOfFinishedOperations, NSUInteger totalNumberOfOperations)) progressBlock
+                                   andOverallCompletionHandler:(void (^)())overallCompletionHandler
+{
+    NSString *cacheKey = @"BatchInsights";
+    NSString *allTimeCacheKey = @"AllTime";
+    NSString *insightsCacheKey = @"Insights";
+    NSString *todayCacheKey = @"Today";
+    
+    NSMutableDictionary *cacheDictionary = [self.ephemory objectForKey:cacheKey];
+    if (cacheDictionary.count == 3) {
```
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/WPStatsService.m]: This one is a little bit hard to follow. Can we replace the `3` with a constant?

**Comment by astralbodies** [comment]
@shaunandrews pointed out the Posts & Views icons were swapped. Switch them back.

**Comment by jleandroperez** [comment]
@astralbodies everything looks great!. Added few minor code comments, please, let me know when it's ready for another round.

Thanks!

File: WordPressCom-Stats-iOS/StatsEphemory.m (hunk 3 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/StatsEphemory.m]: Every time the app would background, the NSCache instance would get automatically evicted. I needed something a little more persistent. I'm not serializing this one to disk (yet) but I plan on doing something like that soon.

Comment (jleandroperez) [inline WordPressCom-Stats-iOS/StatsEphemory.m]: If you do go that way (NSDictionary persistance), take a look at [this helper[(https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPPersistentMutableDictionary.h), might save you some time.

That guy is what Simperium uses, internally, to persist pending changes.

Comment (astralbodies) [inline WordPressCom-Stats-iOS/StatsEphemory.m]: Oooooooooooooooooooo

http://bbsimg.ngfiles.com/1/24316000/ngbbs509a913424500.gif

Comment (jleandroperez) [inline WordPressCom-Stats-iOS/StatsEphemory.m]: It has, internally, a NSCache, but you can just go ahead and replace that with NSMutableDictionary.

Before time, we used to persist changes into a plist file... but that killed the app, when dealing with +10k entities.

File: WordPressCom-Stats-iOS/InsightsTableViewController.m (hunk 2 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/InsightsTableViewController.m]: This will be fixed in a PR soon - I'm revamping the controls and will be making individual pieces tappable.

File: WordPressCom-Stats-iOS/WPStatsService.m (hunk 4 above)
Comment (astralbodies) [inline WordPressCom-Stats-iOS/WPStatsService.m]: I'm going to be rewriting this method completely in another request and will follow the same pattern I did here:

https://github.com/wordpress-mobile/WordPressCom-Stats-iOS/blob/issue/254-insights/WordPressCom-Stats-iOS/WPStatsService.m#L96

**Comment by astralbodies** [comment]
@jancavan @jleandroperez I deployed a new internal StatsDemo binary and I think we're ready for final :+1:

**Comment by isaackeyet** [comment]
I'm going to take a look at this build to get the latest to generate screenshots, but I can also take a look at the progress in the design review.

File: WordPressCom-Stats-iOS/WPStatsService.m (hunk 4 above)
Comment (jleandroperez) [inline WordPressCom-Stats-iOS/WPStatsService.m]: Nice!

**Comment by jleandroperez** [comment]
@astralbodies all looking good on my side. :shipit: !!

**Comment by isaackeyet** [comment]
Design review not complete, no shipit

**Comment by astralbodies** [comment]
Yup @isaackeyet I am waiting :wink:

**Comment by jleandroperez** [comment]
@isaackeyet sorry!! :D. On your mark, code looks great!

**Comment by isaackeyet** [comment]
I've checked out the latest and these are my observations

1. time format seems to use am/pm regardless of region? I get the abbreviation for "pm" in Swedish in this screenshot, but very rarely is time written like this. Elsewhere on this device it's always 24 format, hh:mm.
img_0009_2

2. stuck in loading mode: I'm still able to get the horizontal loading bar to get stuck seemingly indefinitely and sometimes the spinner too.
img_0019_2

3. auto-resizing value labels in the All-Time module do a poor job with large numbers as Jan pointed out. I think it makes sense to wait and fix later, but I also noticed that even rotating doesn't resize the text (which is the first thing I would try if a value can't be fitted inside its container). Probably still not worth it to pursue right now but logging here.

4. tab tap UX: seeing the tabs for "Today's Stats" I expected to be able to tap one and get to the daily view with that filter selected. It turns out the entire bar just takes you there, it doesn't matter where you tap. This was unexpected. It's also weird how this UI is exactly the same, visually, to the one in the overview but in the overview it's contracted to the horizontal navigation but here it's not, resulting in poor margins on smaller devices which are going to break when translated.
img_0011_2

5. there is no five. yay!

Apart from the time format thing I'd be OK with launching this as-is, but having a short-list for an immediate iteration following it. To minimize confusion I think that iteration should be design-first so there are less pain points being brought up only in late-stage design reviews.

cc @jancavan

**Comment by astralbodies** [comment]
> time format seems to use am/pm regardless of region? I get the abbreviation for "pm" in Swedish in this screenshot, but very rarely is time written like this. Elsewhere on this device it's always 24 format, hh:mm.

**Comment by astralbodies** [comment]
> stuck in loading mode: I'm still able to get the horizontal loading bar to get stuck seemingly indefinitely and sometimes the spinner too.

Does is it take switching between the screens for this to happen @isaackeyet?

**Comment by isaackeyet** [comment]
Yes tapping around while stuff is still loading seems to do it.

**Comment by astralbodies** [comment]
> auto-resizing value labels in the All-Time module do a poor job with large numbers as Jan pointed out. I think it makes sense to wait and fix later, but I also noticed that even rotating doesn't resize the text (which is the first thing I would try if a value can't be fitted inside its container). Probably still not worth it to pursue right now but logging here.

This will get fixed before the first release to the App Store :smile: #264 

> tab tap UX: seeing the tabs for "Today's Stats" I expected to be able to tap one and get to the daily view with that filter selected. It turns out the entire bar just takes you there, it doesn't matter where you tap. This was unexpected. It's also weird how this UI is exactly the same, visually, to the one in the overview but in the overview it's contracted to the horizontal navigation but here it's not, resulting in poor margins on smaller devices which are going to break when translated.

Same as above - I will be adding this functionality before it hits the App Store. #261

**Comment by astralbodies** [comment]
@isaackeyet I logged the weird bug with flipping between things as #266

**Comment by astralbodies** [comment]
Thanks everyone for the help reviewing this. At least the next PRs will be much much smaller.

**Comment by isaackeyet** [comment]
Nice. Good job Aaron!
